src/lib/string_pool.rs

```diff
@@ -0,0 +1,506 @@
+use crate::expat_external_h::XML_Char;
+use crate::lib::xmlparse::{ICHAR, ExpatBufRef, ExpatBufRefMut, XmlConvert};
+use crate::lib::xmltok::{ENCODING, XML_CONVERT_INPUT_INCOMPLETE, XML_CONVERT_COMPLETED};
+use crate::stddef_h::size_t;
+
+use bumpalo::Bump;
+use bumpalo::collections::vec::Vec as BumpVec;
+use fallible_collections::FallibleBox;
+use libc::{INT_MAX, c_int, c_uint, c_ulong};
+
+use std::cell::RefCell;
+use std::convert::TryInto;
+use std::mem::swap;
+use std::ptr;
+
+pub const INIT_BLOCK_SIZE: usize = init_block_size_const();
+
+#[cfg(feature = "mozilla")]
+const fn init_block_size_const() -> usize {
+    // FIXME: should be `offset_of(BLOCK, s)`, but that's not supported yet,
+    // so we over-estimate its offset
+    1024 - (std::mem::size_of::<BLOCK>() / std::mem::size_of::<XML_Char>())
+}
+
+#[cfg(not(feature = "mozilla"))]
+const fn init_block_size_const() -> usize {
+    1024
+}
+
+fn poolBytesToAllocateFor(mut blockSize: c_int) -> size_t {
+    /* Unprotected math would be:
+     ** return offsetof(BLOCK, s) + blockSize * sizeof(XML_Char);
+     **
+     ** Detect overflow, avoiding _signed_ overflow undefined behavior
+     ** For a + b * c we check b * c in isolation first, so that addition of a
+     ** on top has no chance of making us accept a small non-negative number
+     */
+    let stretch: size_t = ::std::mem::size_of::<XML_Char>() as c_ulong; /* can be 4 bytes */
+    if blockSize <= 0 {
+        return 0u64;
+    }
+    if blockSize > (INT_MAX as c_ulong).wrapping_div(stretch) as c_int {
+        return 0u64;
+    }
+    let stretchedBlockSize: c_int = blockSize * stretch as c_int;
+    let bytesToAllocate: c_int =
+        (12u64).wrapping_add(stretchedBlockSize as c_uint as c_ulong) as c_int;
+    if bytesToAllocate < 0 {
+        return 0u64;
+    }
+    bytesToAllocate as size_t
+}
+
+rental! {
+    mod rental_pool {
+        use super::*;
+
+        #[rental(debug)]
+        pub(crate) struct InnerStringPool {
+            bump: Box<Bump>,
+            current_bump_vec: RefCell<BumpVec<'bump, XML_Char>>,
+        }
+    }
+}
+
+use rental_pool::InnerStringPool;
+
+pub(crate) struct StringPool(Option<InnerStringPool>);
+
+impl StringPool {
+    pub(crate) fn try_new() -> Result<Self, ()> {
+        let bump = Bump::try_new().map_err(|_| ())?;
+        let boxed_bump = Box::try_new(bump).map_err(|_| ())?;
+
+        Ok(StringPool(Some(InnerStringPool::new(
+            boxed_bump,
+            |bump| RefCell::new(BumpVec::new_in(&bump)),
+        ))))
+    }
+
+    fn inner(&self) -> &InnerStringPool {
+        self.0.as_ref().unwrap()
```

Comment:
What runtime check are you concerned about? The unwrap? We could do `self.0.as_ref().unwrap_or_else(|| unsafe { unreachable_unchecked() });` so that it gets optimized out. The only time it is `None` is in the middle of `clear()`

Reply:
Yeah, the unwrap. I'm not sure that version is any better than the UB from uninit above.

Reply:
I think `unreachable_unchecked` is easier to reason about. For example, we know that there's only one location (`clear`) that it cannot be called from. Writing to/from uninit data involving references is largely a grey area that could possible change over time as Rust's model formalizes.